subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+	if (!fs->sblock) {
+		LOG_ERR("File system data not found. Panic...");
+		k_panic();
+	}
+
+	switch (EXT2_DATA_SBLOCK(fs)->s_errors) {
+	case EXT2_ERRORS_CONTINUE:
+		/* Do nothing */
+		break;
+	case EXT2_ERRORS_RO:
+		LOG_WRN("Marking file system as read only");
+		fs->flags |= EXT2_DATA_FLAGS_RO;
+		break;
+	case EXT2_ERRORS_PANIC:
+		LOG_ERR("Panic...");
+		k_panic();
+		break;
+	default:
+		LOG_ERR("Unrecognized errors behavior in superblock s_errors field. Panic...");
+		k_panic();
+	}
+}
+
+void *ext2_heap_alloc(size_t size)
+{
+	void *ptr = k_heap_alloc(&ext2_heap, size, K_NO_WAIT);
+
+	if (!ptr) {
+		return NULL;
+	}
+
+	memset(ptr, 0, size);
+	return ptr;
+}
+
+void  ext2_heap_free(void *ptr)
+{
+	k_heap_free(&ext2_heap, ptr);
+}
+
+/* Block operations --------------------------------------------------------- */
+
+struct ext2_block *ext2_get_block(struct ext2_data *fs, uint32_t block)
+{
+	int ret;
+	struct ext2_block *b;
+
+	ret = k_mem_slab_alloc(&ext2_block_struct_slab, (void **)&b, K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("get block: alloc block struct error %d", ret);
+		return NULL;
+	}
+
+	ret = k_mem_slab_alloc(&ext2_block_memory_slab, (void **)&b->data, K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("get block: alloc block memory error %d", ret);
+		return NULL;
+	}
+
+	b->num = block;
+	b->flags = EXT2_BLOCK_ASSIGNED;
+	ret = fs->read_block(fs, b->data, block);
+	if (ret < 0) {
+		LOG_ERR("get block: read block error %d", ret);
+		return NULL;
+	}
+
+	return b;
+}
+
+struct ext2_block *ext2_get_empty_block(struct ext2_data *fs)
+{
+	int ret;
+	struct ext2_block *b;
+
+	ret = k_mem_slab_alloc(&ext2_block_struct_slab, (void **)&b, K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("get empty block: alloc block struct error %d", ret);
+		return NULL;
+	}
+
+	ret = k_mem_slab_alloc(&ext2_block_memory_slab, (void **)&b->data, K_NO_WAIT);
+	if (ret < 0) {
+		LOG_ERR("get empty block: alloc block error %d", ret);
+		return NULL;
+	}
+
+	b->num = 0;
+	b->flags = 0;
+	memset(b->data, 0, fs->block_size);
+
+	return b;
+}
+
+int ext2_write_block(struct ext2_data *fs, struct ext2_block *b)
+{
+	int ret;
+
+	if (!(b->flags & EXT2_BLOCK_ASSIGNED)) {
+		return -EINVAL;
+	}
+
+	ret = fs->write_block(fs, b->data, b->num);
+	if (ret < 0) {
+		return ret;
+	}
+	return 0;
+}
+
+void ext2_drop_block(struct ext2_block *b)
+{
+	if (b == NULL) {
+		return;
+	}
+
+	if (b != NULL && b->data != NULL) {
+		k_mem_slab_free(&ext2_block_memory_slab, (void **)&b->data);
+		k_mem_slab_free(&ext2_block_struct_slab, (void **)&b);
+	}
+}
+
+int ext2_init_blocks_slab(struct ext2_data *fs)
+{
+	int ret;
+	uint32_t num_blocks = CONFIG_EXT2_BLOCKS_BUF_SIZE / fs->block_size;
+
+	memset(__ext2_block_memory_buffer, 0, CONFIG_EXT2_BLOCKS_BUF_SIZE);
+	memset(__ext2_block_struct_buffer, 0, CONFIG_EXT2_BLOCKS_BUF_SIZE / 100);
+
+	ret = k_mem_slab_init(&ext2_block_struct_slab, __ext2_block_struct_buffer,
+			sizeof(struct ext2_block), num_blocks);
+	if (ret < 0) {
```

Comment:
Here or in caller, caller would probably be better.

Reply:
> I will add logs here. I think it can't be done in caller (because we can't determine which function call failed because they may return identical errors).

Yeah (said facepalming over his second comment).